### PR TITLE
Dynamic reload of consul agent name

### PR DESCRIPTION
**Pull request by clems4ever**
This commit allows to dynamically rename consul agents in client mode only using consul reload.

When a reload with name changed is triggered, the agent leaves the serf cluster and join it back with its new name using a subset of IPs of the previously known peers as start addresses. This commit does not handle server being renamed to mitigate possible risks on the cluster.

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -3514,19 +3541,26 @@ func (a *Agent) ReloadConfig() error {
 	// breaking some existing behavior.
 	newCfg.NodeID = a.config.NodeID
 
+	nodeNameChanged := false
```
Comment (clems4ever) [inline agent/agent.go:3544]: Do nothing in server mode.

Comment (clems4ever) [reply agent/agent.go:3544]: Add warning message in the case of server.

File: agent/agent.go [hunk 2]
```diff
@@ -3613,6 +3655,23 @@ func (a *Agent) reloadConfigInternal(newCfg *config.RuntimeConfig) error {
 		return err
 	}
 
+	if nodeNameChanged {
+		a.logger.Debug("Node name has been modified, leave and join with the new name is requested")
```
Comment (pierresouchay) [inline agent/agent.go:3659]: I think it deserves a WARN or INFO level at the minimum

File: agent/agent.go [hunk 3]
```diff
@@ -3603,6 +3637,14 @@ func (a *Agent) reloadConfigInternal(newCfg *config.RuntimeConfig) error {
 		return err
 	}
 
+	_, isServer := a.delegate.(*consul.Server)
+	nodeNameChanged := !isServer && a.config.NodeName != newCfg.NodeName
```
Comment (pierresouchay) [inline agent/agent.go:3641]: You might add a comment here explaining why you don't want to rename servers, even better:
if node is changed, but isServer: add a message explaining that nodeRenaming is not supported on server mode and requires to leave cluster first and restart server.

**Review by pierresouchay — APPROVED** [review]
Sounds pretty cool!
A few minor comments, but pretty sure HashiCorp will have many other ones

File: agent/agent.go (hunk 3 above)
Comment (clems4ever) [reply agent/agent.go:3641]: Already done further down in the code. I'm reworking the conditions so that it's clearer.

File: agent/agent.go (hunk 2 above)
Comment (clems4ever) [reply agent/agent.go:3659]: Sure, it's gonna be warn then.

**Pull request closed without merging**